ford/md_extensions.py

```diff
@@ -0,0 +1,307 @@
+            end_idx: end line index, one of: @end..., empty line,
+                start marker of next  admonition, last line of text.
+        """
+
+        type: str
+        start_idx: int
+        end_idx: Optional[int] = None
+
+    def run(self, lines: List[str]) -> List[str]:
+        self.lines = lines
+        self._find_admonitions()
+        self._process_admonitions()
+        return self.lines
+
+    def _find_admonitions(self):
+        """Scans the lines to search for admonitions."""
+        self.admonitions = []
+        current_admonition = None
+
+        for idx, line in enumerate(self.lines):
+            match = self.ADMONITION_RE.search(line)
+
+            if match and not match.group("end"):
+                if current_admonition:
+                    if not current_admonition.end_idx:
+                        current_admonition.end_idx = idx
+                    self.admonitions.append(current_admonition)
+                current_admonition = self.Admonition(
+                    type=match.group("type"),
+                    start_idx=idx,
+                )
+
+            elif match and match.group("end"):
+                if current_admonition:
+                    if match.group("type").lower() != current_admonition.type.lower():
+                        pass  # error: type of start and end marker dont' match
+                    current_admonition.end_idx = idx
+                    self.admonitions.append(current_admonition)
+                    current_admonition = None
+                else:
+                    pass  # error: end marker found without start marker
+
+            elif line == "" and current_admonition and not current_admonition.end_idx:
+                # empty line encountered while in an admonition. We set end_line but don't
+                # move it to the list yet since an end marker (@end...) may follow
+                # later.
+                current_admonition.end_idx = idx
+
+        if current_admonition:
+            # We reached the last line and the last admonition wasn't moved to the list yet.
+            if not current_admonition.end_idx:
+                current_admonition.end_idx = idx
+            self.admonitions.append(current_admonition)
+
+    def _process_admonitions(self):
+        """Processes the admonitions to convert the lines to the markdown syntax."""
+
+        # We handle the admonitions in the reverse order since
+        # we may be deleting lines.
+        for admonition in self.admonitions[::-1]:
+
+            # last line--deal with possible text before or after end marker
+            idx = admonition.end_idx
+            m = self.ADMONITION_RE.search(self.lines[idx])
+            if m:
+                if m.group("posttxt"):
+                    self.lines.insert(idx + 1, m.group("posttxt"))
+                del self.lines[idx]
+                if m.group("pretxt"):
+                    self.lines.insert(idx, " " * self.INDENT_SIZE + m.group("pretxt"))
```

Comment:
Maybe store `INDENT = " " * self.INDENT_SIZE` to avoid this duplicated expression?

It might also be worth considering using [`textwrap.indent`](https://docs.python.org/3/library/textwrap.html#textwrap.indent) to make this a bit more explicit?

We should probably try to preserve the existing indentation, as with this current implementation notes in lists etc are not indented, and numbered lists are restarted after the note:

![image](https://user-images.githubusercontent.com/1486942/165756078-d6ed5dfd-5cd0-431e-aaf6-74182f769db7.png)

rather than:

![image](https://user-images.githubusercontent.com/1486942/165756311-744f9106-e7fe-44d2-9758-812e8e34928b.png)